libtiledbvcf/src/stats/variant_stats_reader.cc

```diff
@@ -45,13 +46,230 @@ inline int pos_comparator(const void* a, const void* b) {
   }
 }
 
+bool AFMap::RefBlockComp::operator()(
+    const RefBlock& a, const RefBlock& b) const {
+  if (a.start < b.start) {
+    return true;
+  } else {
+    if (a.start == b.start && a.end < b.end) {
+      return true;
+    }
+  }
+  return false;
+}
+
+bool AFMap::RefBlockComp::operator()(
+    const RefBlock* a, const RefBlock* b) const {
+  if (a->end < b->end) {
+    return true;
+  } else {
+    if (a->end == b->end && a->start < b->start) {
+      return true;
+    }
+  }
+  return false;
+}
+
+inline void AFMap::advance_to_ref_block(uint32_t pos) {
+  if (pos + 1 == active_pos_) {
+    return;
+  }
+  if (pos + 1 < active_pos_) {
+    throw std::runtime_error(
+        "[VariantStatsReader] ref block computation performed on incompletely "
+        "sorted positions");
+  }
+  active_pos_++;
+  // ref blocks entering scope
+  while (selected_ref_block_ != ref_block_cache_.end() &&
+         selected_ref_block_->start <= pos) {
+    ac_sum_ += selected_ref_block_->ac;
+    an_sum_ += selected_ref_block_->an;
+    selected_ref_block_++;
+  }
+  // ref blocks exiting scope
+  while (selected_ref_block_end_ != ref_block_by_end_.end() &&
+         (**selected_ref_block_end_).end < pos) {
+    ac_sum_ -= (**selected_ref_block_end_).ac;
+    an_sum_ -= (**selected_ref_block_end_).an;
+    selected_ref_block_end_++;
+  }
+}
+
+std::tuple<float, uint32_t, uint32_t> AFMap::af(
+    uint32_t pos, const std::string& allele) {
+  // calculate af = ac / an
+  std::unordered_map<
+      uint32_t,
+      std::pair<int, std::unordered_map<std::string, int>>>::const_iterator
+      pos_map_iterator = ac_map_.find(pos);
+
+  // Return -1.0 if the allele was not called
+  if (pos_map_iterator == ac_map_.end()) {
+    return {-1.0, 0, 0};
+  }
+
+  const auto pos_map = pos_map_iterator->second;
+
+  // We don't know that allele was called in this sample. Ask nicely for the
+  // allele count.
+  auto next_allele = pos_map.second.find(allele);
+
+  // First multiply by 1.0 to force a float type, then look up AC from the
+  // above iterator. Substitute 0 for the AC value if it is absent in pos_map.
+  // Divide by AN (pos_map.first) to get AF.
+  if (next_allele == pos_map.second.end()) {
+    return {0, 0, pos_map.first};
+  }
+  return {
+      1.0 * next_allele->second / pos_map.first,
+      next_allele->second,
+      pos_map.first};
+}
+
+std::tuple<float, uint32_t, uint32_t> AFMap::af(
```

Comment:
This version of `af` looks a lot like the previous version of `af`. Can they be combined?

Maybe set the default value of `num_samples` to 0 and use that to decide what to return?

Reply:
There were two variants of `af` already; that exploded to four. Distinguishing between fixed vs. variable samples was already present, and was done this way because the method is small, in a header, and called frequently, whereby branching would have a performance impact.